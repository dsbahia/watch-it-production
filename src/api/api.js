import axios from "axios";

async function makeRequest(url, params) {
  try {
    const response = await axios.get(url, {
      params,
      headers: {
        Authorization: `Bearer ${process.env.REACT_APP_API_KEY}`,
      },
    });
    return response.data;
  } catch (error) {
    return null;
  }
}

async function searchMovies(query) {
  const apiUrl = "https://api.themoviedb.org/3/search/movie";
  const params = {
    query,
    language: "en-GB",
    region: "GB",
  };
  return makeRequest(apiUrl, params);
}

async function trendingMovies() {
  const apiUrl = "https://api.themoviedb.org/3/trending/movie/day";
  const params = {
    language: "en-GB",
  };
  return makeRequest(apiUrl, params);
}

async function trendingTVShows() {
  const apiUrl = "https://api.themoviedb.org/3/trending/tv/day";
  const params = {
    language: "en-GB",
  };
  return makeRequest(apiUrl, params);
}

async function searchMovieById(type, id) {
  const apiUrl = `https://api.themoviedb.org/3/${type}/${id}`;
  const params = {
    language: "en-GB",
  };
  return makeRequest(apiUrl, params);
}

async function topRatedMovies() {
  const apiUrl = "https://api.themoviedb.org/3/movie/top_rated";
  const params = {
    language: "en-GB",
  };
  return makeRequest(apiUrl, params);
}

async function topRatedTvShows() {
  const apiUrl = "https://api.themoviedb.org/3/tv/top_rated";
  const params = {
    language: "en-GB",
  };
  return makeRequest(apiUrl, params);
}

async function upcomingMovies() {
  const apiUrl = "https://api.themoviedb.org/3/movie/upcoming";
  const params = {
    language: "en-GB",
  };
  return makeRequest(apiUrl, params);
}

async function airingTvShows() {
  const currentDate = new Date();
  const futureDate = new Date();

  futureDate.setDate(currentDate.getDate() + 1);

  const formattedCurrentDate = currentDate.toISOString().split("T")[0];
  const formattedFutureDate = futureDate.toISOString().split("T")[0];

  const apiUrl = "https://api.themoviedb.org/3/tv/on_the_air";
  const params = {
    "air_date.gte": formattedCurrentDate,
    "air_date.lte": formattedFutureDate,
    include_adult: "false",
    include_null_first_air_dates: "false",
    language: "en-US",
    sort_by: "popularity.desc",
    with_origin_country: "GB",
  };
  return makeRequest(apiUrl, params);
}
async function movieTrailer(movieId) {
  const apiUrl = `https://api.themoviedb.org/3/movie/${movieId}/videos`;
  const params = {
    language: "en-GB",
  };
  return makeRequest(apiUrl, params);
}

async function tvTrailer(tvId) {
  const apiUrl = `https://api.themoviedb.org/3/tv/${tvId}/videos`;
  const params = {
    language: "en-GB",
  };
  return makeRequest(apiUrl, params);
}

export default {
  makeRequest,
  searchMovies,
  trendingMovies,
  trendingTVShows,
  searchMovieById,
  topRatedMovies,
  topRatedTvShows,
  upcomingMovies,
  airingTvShows,
  movieTrailer,
  tvTrailer,
};
