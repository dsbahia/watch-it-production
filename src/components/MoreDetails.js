import React, { useState, useEffect } from "react";
import toast from "react-hot-toast";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faStar } from "@fortawesome/free-solid-svg-icons";
import "../styles/moredetails.css";
import api from "../api/api";

function MoreDetails({ type, id }) {
  const [moviesData, setMoviesData] = useState(false);

  useEffect(() => {
    async function fetchMovieById() {
      try {
        const data = await api.searchMovieById(type, id);
        setMoviesData(data);
      } catch (error) {
        const errorMsg = "An error occurred. Please try again later.";
        toast.error(errorMsg, {
          duration: 5000,
        });
      }
    }
    fetchMovieById();
  }, [type, id]);

  if (moviesData === null) {
    return <div>Loading...</div>;
  }

  const genresList = moviesData.genres
    ? moviesData.genres.map((genre) => <div key={genre.id}>{genre.name}</div>)
    : [];
  const lastEpisodeRuntime = moviesData.last_episode_to_air
    ? `${moviesData.last_episode_to_air.runtime} Minutes`
    : `${moviesData.runtime} minutes`;
  const budgetCheck =
    moviesData.budget != null
      ? `Budget: $ ${moviesData.budget.toLocaleString()}`
      : "Budget information not available";
  const revenueCheck =
    moviesData.revenue != null
      ? `Revenue: $ ${moviesData.revenue.toLocaleString()}`
      : "Revenue information not available";
  const releaseDateCheck = moviesData.first_air_date
    ? `First Episode Release Date: ${moviesData.first_air_date}`
    : `Release  Date: ${moviesData.release_date}`;

  function renderRatingStars(rating) {
    const maxRating = 10;
    const fullStars = Math.floor(rating);
    const halfStar = rating - fullStars >= 0.5;
    const emptyStars = maxRating - fullStars - (halfStar ? 1 : 0);

    const starIcons = Array.from(
      { length: fullStars + (halfStar ? 1 : 0) + emptyStars },
      (_, index) => {
        if (index < fullStars) {
          return (
            <FontAwesomeIcon
              key={index}
              icon={faStar}
              style={{ color: "green" }}
            />
          );
        }
        if (index === fullStars && halfStar) {
          return (
            <FontAwesomeIcon
              key="half"
              icon={faStar}
              style={{ color: "yellow" }}
            />
          );
        }
        return (
          <FontAwesomeIcon
            key={index}
            icon={faStar}
            style={{ color: "gray" }}
          />
        );
      },
    );

    return starIcons;
  }

  const rating = Math.round(moviesData.vote_average * 10) / 10;
  const ratingStars = renderRatingStars(rating);

  return (
    <div className="more-details-container">
      <div className="more-details-overview">{moviesData.overview}</div>
      <div className="more-details-item">{budgetCheck}</div>
      <div className="more-details-item">{revenueCheck}</div>
      <div className="more-details-item">
        Genres:
        <div>{genresList}</div>
      </div>
      <div className="more-details-item">{releaseDateCheck}</div>
      <div className="more-details-item">Rating: {ratingStars}</div>
      <div className="more-details-item">Runtime: {lastEpisodeRuntime}</div>
    </div>
  );
}

export default MoreDetails;
