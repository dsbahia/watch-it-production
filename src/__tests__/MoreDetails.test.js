import React from "react";
import { render, waitFor, screen } from "@testing-library/react";
import MoreDetails from "../components/MoreDetails";

jest.mock("react-hot-toast", () => ({
  error: jest.fn(),
}));

jest.mock("../api/api", () => ({
  searchMovieById: jest.fn(() =>
    Promise.resolve({
      overview: "Mock Movie Overview",
      budget: 10000000,
      revenue: 50000000,
      genres: [
        { id: 1, name: "Action" },
        { id: 2, name: "Adventure" },
      ],
      release_date: "2023-09-22",
      vote_average: 7.5,
      last_episode_to_air: { runtime: 60 },
    }),
  ),
}));
describe("MoreDetails", () => {
  const { asFragment } = render(<MoreDetails type="movie" id={123} />);

  it("renders  correctly", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders data", async () => {
    const mockMovieData = {
      overview: "Mock Movie Overview",
      budget: 10000000,
      revenue: 50000000,
      genres: [
        { id: 1, name: "Action" },
        { id: 2, name: "Adventure" },
      ],
      release_date: "2023-09-22",
      vote_average: 7.5,
      last_episode_to_air: { runtime: 60 },
    };
    jest.mock("../api/api", () => ({
      searchMovieById: jest.fn(() =>
        Promise.resolve({
          overview: "Mock Movie Overview",
          budget: 10000000,
          revenue: 50000000,
          genres: [
            { id: 1, name: "Action" },
            { id: 2, name: "Adventure" },
          ],
          release_date: "2023-09-22",
          vote_average: 7.5,
          last_episode_to_air: { runtime: 60 },
        }),
      ),
    }));

    require("../api/api").searchMovieById.mockResolvedValueOnce(mockMovieData);

    render(<MoreDetails type="movie" id={1} />);

    await waitFor(() => {
      const overviewElement = screen.getByText("Mock Movie Overview");
      const budgetElement = screen.getByText("Budget: $ 10,000,000");
      const revenueElement = screen.getByText("Revenue: $ 50,000,000");
      const genresElement = screen.getByText("Genres:");
      const runtimeElement = screen.getByText("Runtime: 60 Minutes");

      expect(overviewElement).toBeInTheDocument();
      expect(budgetElement).toBeInTheDocument();
      expect(revenueElement).toBeInTheDocument();
      expect(genresElement).toBeInTheDocument();
      expect(runtimeElement).toBeInTheDocument();
    });
  });
});
