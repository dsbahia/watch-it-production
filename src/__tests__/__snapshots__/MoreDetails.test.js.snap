// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoreDetails renders  correctly 1`] = `
<DocumentFragment>
  <div
    class="more-details-container"
  >
    <div
      class="more-details-overview"
    />
    <div
      class="more-details-item"
    >
      Budget information not available
    </div>
    <div
      class="more-details-item"
    >
      Revenue information not available
    </div>
    <div
      class="more-details-item"
    >
      Genres:
      <div />
    </div>
    <div
      class="more-details-item"
    >
      Release  Date: undefined
    </div>
    <div
      class="more-details-item"
    >
      Rating: 
    </div>
    <div
      class="more-details-item"
    >
      Runtime: undefined minutes
    </div>
  </div>
</DocumentFragment>
`;
